added SuperFlappyBirds scriptmodule to libretrocores

IgniparousTempest opened the pull request:
Super Flappy Birds is an original-ish game created by me, specifically with retropie in mind. The game and screenshots are [here](https://github.com/IgniparousTempest/libretro-superflappybirds).

It is written as a libretro core, hence why the script is under the `libretrocores` folder, but it installs to the ports folder on retropie. This pattern seems odd, but it is what Dinothawr does, see [here](https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/libretrocores/lr-dinothawr.sh).

hhromic commented on scriptmodules/libretrocores/lr-superflappybirds.sh:
```diff
@@ -0,0 +1,45 @@
+function build_lr-superflappybirds() {
+    cmake .
+    make
+    md_ret_require="$md_build/liblr_superflappybirds.so"
```
Normally, libretro cores are named `*_libretro.so`, perhaps you should align your project to this convention too, e.g. `superflappybirds_libretro.so`.

IgniparousTempest replied:
Done

hhromic commented on scriptmodules/libretrocores/lr-superflappybirds.sh:
```diff
@@ -0,0 +1,45 @@
+
+function install_lr-superflappybirds() {
+    md_ret_files=(
+        'liblr_superflappybirds.so'
```
Same here.

IgniparousTempest replied:
Done

hhromic commented on scriptmodules/libretrocores/lr-superflappybirds.sh:
```diff
@@ -0,0 +1,45 @@
+function configure_lr-superflappybirds() {
+    setConfigRoot "ports"
+
+    addPort "$md_id" "superflappybirds" "Super Flappy Birds" "$md_inst/liblr_superflappybirds.so" "$romdir/ports/superflappybirds/"
```
Same here.

IgniparousTempest replied:
Done

IgniparousTempest commented:
@hhromic I have fixed the library name as you suggested.

The `$romdir/ports/superflappybirds/` is being used to store high scores.

I wasn't sure how else to do it, as `RETRO_ENVIRONMENT_GET_SAVE_DIRECTORY` was returning `/` as the save directory.

hhromic commented:
@IgniparousTempest thanks for the changes. Just for your information, the library naming convention is more from the libretro project, than from RetroPie :)

Regarding the high scores saving, I think most cores use the `system` dir for this purpose and not the `save` one. The latter I understand is for game statesaves/battery backups. So maybe you want to explore that in the libretro API.  In RetroPie, the system dir is mapped to `$biosdir`.

```
#define RETRO_ENVIRONMENT_GET_SYSTEM_DIRECTORY 9
                                           /* const char ** --
                                            * Returns the "system" directory of the frontend.
                                            * This directory can be used to store system specific 
                                            * content such as BIOSes, configuration data, etc.
                                            * The returned value can be NULL.
                                            * If so, no such directory is defined,
                                            * and it's up to the implementation to find a suitable directory.
                                            *
                                            * NOTE: Some cores used this folder also for "save" data such as 
                                            * memory cards, etc, for lack of a better place to put it.
                                            * This is now discouraged, and if possible, cores should try to 
                                            * use the new GET_SAVE_DIRECTORY.
                                            */
```

You mention that your game has content (game files), but your scriptmodule doesn't seem to be installing any of that. Is the user supposed to copy the game content manually herself? You should document this in the `rp_module_help` variable.

I saw that your project has the game files in github already, maybe you can sort out a way to automatically download these into the `$romdir`. The idea is that the user, after installing your game, shouldn't need to do any extra steps but to launch it from EmulationStation.

hhromic commented:
Ah, my bad! Just realised you are copying the `resources` folder in the install function. This is being copied to `/opt/retropie/libretrocores/lr-superflappybirds`.
So there is no real use for a directory in the `ports` system as everything in your game is included in the game installation folder. You just need the standard shell script in `$romdir/ports`.

Perhaps you can simply leave out the last parameter of `addPort` then and should be good to go.

IgniparousTempest commented:
@hhromic, I have changed my highscore file to use `RETRO_ENVIRONMENT_GET_SYSTEM_DIRECTORY` and have applied your other suggestions. 

Thanks so much for all the help, I really appreciate it.

hhromic commented:
@IgniparousTempest  looking good now!  Happy to help you.

Having your highscores in the `$biosdir` ensures users don't loose them when uninstall/re-installing, and keeps the `$romdir` clean as well.
Don't stop improving your game, is a nice hobby!

All that rests now is for the boss to review and merge your scriptmodule and you are set!

hhromic commented on scriptmodules/libretrocores/lr-superflappybirds.sh:
```diff
@@ -0,0 +1,42 @@
+
+rp_module_id="lr-superflappybirds"
+rp_module_desc="Super Flappy Birds - Multiplayer Flappy Bird Clone"
+rp_module_help="Checkout the wiki on my github page."
```
Just one final remark, can you provide the actual URL to your wiki here? Thanks!

IgniparousTempest replied:
Done

hhromic commented:
It's-a me again :)  I think you also should take a look to the dependencies of your game. For instance, youro build uses `cmake`, so you should add at least the following depends function (usually before the sources function):
```
function depends_lr-superflappybirds() {
    getDepends cmake
}
```
If your game needs any other library to compile, you should add it there too. Cheers!

IgniparousTempest commented:
@hhromic, I updated the url and added the dependency. 

I didn't add cmake, because I thought it came standard on Raspbian. The game's engine/framework was all written by me, so it doesn't have any other dependencies.

hhromic commented:
@IgniparousTempest I tested your scriptmoduel now on my RPI3. Works flawlessly.
Also, I can confirm the entertaining value of your game as I spent a solid 15 minutes with it just for testing!

One final (for real!) suggestion for you to include your license and documentation in the installed files (like many other cores do), by adding this to your install function:
```
function install_lr-superflappybirds() {
    md_ret_files=(
        'superflappybirds_libretro.so'
        'resources'
        'LICENSE'
        'README.md'
    )
}
```

And also check the potential naming licensing issues mentioned in the forum.
@joolswills apart from that, this scriptmodule is ready to go IMHO.

IgniparousTempest commented:
@hhromic I am glad that you enjoyed it. :grin:

I added the licence and the readme as you suggested, that seems like a good idea.

As for the naming issue, i have looked into it and I don't think it is an issue:
1. The article mentioned was from 2014 and I can't find any reference to takedown notices after this point. [TechCrunch](https://techcrunch.com/2014/03/13/company-claiming-rights-to-flappy-trademark-now-sending-notices-to-app-store-game-developers/?guccounter=1)
2. Ultimate Arcade, Inc.'s trademark is now dead, so I doubt they'll be an issue. [Trademark Electronic Search System](http://tmsearch.uspto.gov/bin/showfield?f=doc&state=4807:68eggu.2.15)
3. There are a plethora of **Flappy X** games on the app store, a few released after 2014. [Play Store](https://play.google.com/store/search?q=flappy&c=apps)

I think point 2 is probably the most important there.

joolswills commented:
Thanks.
